README.md
APE - ACE Parsing Engine
========================

Authors: Kaarel Kaljurand, Norbert E. Fuchs, Tobias Kuhn


Introduction
------------

This document explains how APE (ACE Parsing Engine) is compiled and used.

In order to compile and run APE, you first need to install a recent version of SWI-Prolog.
SWI-Prolog is free software and can be downloaded from <http://www.swi-prolog.org>. Note that you
minimally need to install the following SWI Prolog packages: `clib`, `sgml`, and `http`. To view
the documentation embedded in the source files you also need `pldoc`.


Compilation
-----------

Before you can run APE, you have to compile the APE source code. Just execute the file
`make_exe.bat` in the case of Windows or `make install` in the case of Mac OS X, Linux, or any other
Unix system. Both the bat-file and the Makefile are located in the root directory of the APE distribution. As a result (and
given that there were no compilation errors), a new file `ape.exe` is created in the current
directory.

(In some unlikely cases you might have to change the size of the memory areas
used by SWI-Prolog. This is documented at <http://www.swi-prolog.org/pldoc/man?section=memlimit>.)


Execution
---------

APE has to be executed from the command-line. In the command-line terminal, go to the root
directory of APE (where `ape.exe` is located). Then type `ape.exe` in the case of Windows or
`./ape.exe` otherwise. As a result, you should see the following output:

    Attempto Parsing Engine for ACE 6.7, version 6.7-131003
    Copyright 2008-2013, Attempto Group, University of Zurich
    This program comes with ABSOLUTELY NO WARRANTY.
    This is free software, and you are welcome to redistribute it under certain conditions.
    Please visit http://attempto.ifi.uzh.ch for details.
    
    Command-line arguments:
    -text "TEXT"        The input ACE text. If neither -text nor -file is present then the ACE text
                        is read from stdin.
    -file FILENAME      The name or URL of the input file containing the ACE text.
    -ulextext "TEXT"    The user lexicon (taken from a string).
    -ulexfile FILENAME  The user lexicon (taken from a file or URL).
    -solo OUTPUT        Output just one output component. OUTPUT has to be one of {paraphrase,
                        paraphrase1,paraphrase2,owlfss,owlfsspp,owlrdf,owlxml,ruleml,fol,pnf,tptp,
                        tokens,syntax,syntaxpp,syntaxd,syntaxdpp,drs,drsxml,drspp,drshtml}.
    -cdrs               Output the DRS as a Prolog term.
    -cdrsxml            Output the DRS in XML.
    -cdrspp             Output the DRS in pretty-printed form in plain text.
    -cdrshtml           Output the DRS in pretty-printed form in HTML.
    -cparaphrase        Output a paraphrase which is a "best-effort" combination of paraphrase1 and
                        paraphrase2.
    -cparaphrase1       Output a paraphrase which uses full sentences instead of relative clauses.
    -cparaphrase2       Output a paraphrase which uses relative clauses instead of full sentences.
    -ctokens            Output tokens as a Prolog list of lists.
    -csentences         Output sentences as a Prolog list.
    -csyntax            Output simplified syntax trees as a Prolog list.
    -csyntaxpp          Output simplified syntax trees in pretty-printed form.
    -csyntaxd           Output plain syntax trees as a Prolog list (for debugging).
    -csyntaxdpp         Output plain syntax trees in pretty-printed form (for debugging).
    -cowlfss            Output OWL/SWRL in the Functional-Style Syntax representation (as Prolog
                        term).
    -cowlfsspp          Output OWL/SWRL in the Functional-Style Syntax representation (pretty-
                        printed).
    -cowlxml            Output OWL/SWRL in the XML representation.
    -cowlrdf            Output OWL/SWRL in the RDF/XML representation. DEPRECATED
    -cruleml            Output RuleML representation of the DRS.
    -cfol               Output standard first-order logic representations (default form) of the DRS
                        as a Prolog term.
    -cpnf               Output standard first-order logic representations (prenex normal form) of
                        the DRS as a Prolog term.
    -ctptp              Output TPTP representation of the DRS.
    -uri URI            URI for the OWL outputs.
    -noclex             Ignore the lexicon entries that are compiled into the executable.
    -guess              Guess the word-class of unknown words.
    -server             Launch a socket interface to APE at port 2766 (0xACE).
    -httpserver         Launch an HTTP interface to APE at port 8000.
    -port NUMBER        Override the default port of either the socket or the HTTP interface.
    -version            Shows version information.
    -help               Shows this help page.

APE can be used via four different interfaces:

- command-line interface
- socket interface
- HTTP interface (webservice)
- directly from Prolog or Java

All these possibilities are described in the following sections.


### Command-line interface to APE

The following command parses the text "John waits." and outputs the DRS in XML representation and
the syntax tree:

    ./ape.exe -text "John waits." -cdrsxml -csyntax

In the case of Windows, you have to omit the first two characters `./`. The next example parses the
text that is inside of the file `ace.txt` and outputs the OWL FSS representation:

    ./ape.exe -file ace.txt -solo owlfss

If you omit both arguments, `text` and `file`, the ACE text is read from the standard input:

    echo "Every mammal is an animal." | ./ape.exe -solo drspp

Note that this does not work under Windows.

If you just execute the line

    ./ape.exe -solo drspp

then the terminal waits for an input. In this case, you can type your ACE text into the terminal
window. Once you have done so, press `Enter` and `Ctrl-D` to tell the terminal that you are
finished. The output (the pretty printed DRS in our case) is then shown below the ACE text you just
entered. Again, this does not work under Windows.


### Socket interface to APE

A socket interface to APE is started by giving the argument `-server` to APE. The default port,
2766 (0xACE), can be overridden by specifying a different port number as a value to the argument
`-port`.

A good way to start the server on a Unix command-line (e.g. _bash_) is:

    nohup swipl -x ape.exe -- -server -port 3453 > stdout.txt 2> stderr.txt &

On Mac OS X, one could use _launchctl_ instead.

The socket interface knows only one command:

    get(Parameters).

where `Parameters` is a list of parameters accepted by `get_ape_results/2`. Note the dot at the end
of the command! (There should also be a newline after the dot.) Given the input, the server runs
`get_ape_results/2` on it and sends back the results, followed by
`APESERVERSTREAMEND` on a separate line. If the input command is syntactically incorrect then an
error message is logged into STDERR. In any case, the connection to the client is closed.

Examples of input commands:

    get([text='Every man is a human.', cparaphrase1=on]).
    get([text='Every man is a human.', solo=paraphrase1]).
    get([text='Every man is a a human.', cparaphrase1=on]).
    get([text='Every man is a human.', cinput=on, cdrs=on, cparaphrase=on, ctokens=on, csyntax=on]).

Session example, assuming that the APE socket server listens at port 3453:

    $ telnet 127.0.0.1 3453
    Trying 127.0.0.1...
    Connected to localhost.
    Escape character is '^]'.
    get([text='Every man is a human.', solo=paraphrase1]).
    If there is a man X1 then the man X1 is a human.
    APESERVERSTREAMEND
    Connection closed by foreign host.


### HTTP interface to APE

An HTTP interface to APE is started by giving the argument `-httpserver` to APE. The default port,
8000, can be overridden by specifying a different port number as a value to the argument `-port`.

A good way to start the server on a Unix command-line (e.g. _bash_) is:

    nohup swipl -x ape.exe -- -httpserver -port 8001 > stdout.txt 2> stderr.txt &

On Mac OS X, one could use _launchctl_ instead.

In case there is an error message in the error output, saying that

    ERROR: socket_error(Address already in use)

then try another port number. For example, one would get this error when running the above command
twice: during the first time a server is started which starts to listen requests at port 8001,
during the second time a second server is started but since the port is already being used by the
first server, there is a conflict and the second server crashes.

You can test the server by loading the following URL in your browser.

    http://localhost:8001/?text=Every+man+is+a+human.&solo=drshtml

The result should be an HTML-rendering of the DRS of the sentence "Every man is a human.".

The complete description of this webservice interface as well as some example clients are available
at http://attempto.ifi.uzh.ch/site/docs/ape_webservice.html. (Note though that the parameter
`ulexreload` is not supported by this HTTP interface.)

Note that the parameter names and values, and their meaning is the same as for the command-line
client. Also the results' format is the same. The only difference is that for security reasons the
webservice cannot access local files, i.e. the input ACE text can be passed as a string or via a
pointer an HTTP resource (i.e. URL) but not via pointing to a local file (such as
`/var/acetexts/sometext.ace.txt`).


### Using APE from Java programs

See the documentation in [java/](java/).


### Some examples

ACE text from URL

    ./ape.exe -file http://attempto.ifi.uzh.ch/site/acetexts/example1.ace.txt -solo owlxml

Lexicon from string

    ./ape.exe -text "Every mman is a hhuman." -ulextext "noun_sg(mman, mman, masc). noun_sg(hhuman, hhuman, neutr)." -cparaphrase1

Lexicon from a file

    echo "noun_sg(mman, mman, masc). noun_sg(hhuman, hhuman, neutr)." > ulex.pl
    ./ape.exe -text "Every mman is a hhuman." -ulexfile ulex.pl -cparaphrase1


CHANGES.md
APE Change Log
==============

Version 6.7-180714
------------------

- Package APE as a [SWI-Prolog extension package](http://www.swi-prolog.org/howto/Pack.txt)
  (Thanks to @fnogatz, see more at https://github.com/Attempto/APE/pull/24)
- Small fixes to improve compatibility with dependencies

Version 6.7-131003
------------------

This release implements two new ACE features and brings minor improvements.

- New in ACE 6.7: Support for the exponentiation operator ^ in arithmetical expressions,
  e.g. "2^3^4", which is parsed as (2^3)^4.
- New in ACE 6.7: Support for how many/how much queries, which ask for numbers of
  countable nouns or for amounts of mass or measurement nouns,
  e.g. "How many beaches have how much sand?"
- APE can now translate ACE texts into the RuleML format.
- The coverage of the ACE-to-OWL translator has been improved.
- The dependencies of the Java Interface for APE on external libraries are now managed using Maven.
- The large Clex lexicon (~100,000 entries) was migrated to a separate repository [Attempto/Clex](https://github.com/Attempto/Clex).
  As before, the APE source distribution contains a reduced (~2000 entries) version of this lexicon.


Version 6.6-110816
------------------

This is a bugfix release that brings minor changes and updates:

- ACE to DRS translation:
  - fixed: sentences like "For every woman of who everybody waits." now correctly fail
  - fixed: a bug that in rare cases generated a DRS with duplicate referents
  - fixed: duplicate DRS conditions are now always suppressed
  - fixed: misguiding error message for imperatives
  - deprecated: the construct "the ... of which/who"
- DRS to FOL translation:
  - fixed: disjunctions related by anaphoric references (e.g. the DRS of "A man sleeps or the man
    waits.") are now correctly mapped to FOL
- DRS to ACE translation:
  - improved: handling of disjuntions, e.g. "Every dog barks or a cat sleeps." is now paraphrased
    as "A cat sleeps or if there is a dog X1 then the dog X1 barks."
  - restored: support for 0, e.g. "There are 0 men."
- DRS to OWL/SWRL translation:
  - the pretty-printed output (owlfsspp) now conforms to the OWL Functional-Style Syntax (and can
    be parsed by the OWL-API)
  - avoid including redundant owl:Thing in intersections
  - now supported: the DRS of "John does not see what?"
  - fixed: sentences like "There is X. If X is a fireman then X is a man." now map to OWL (instead
    of SWRL as before)
  - improved: the SWRL-mapping now supports formula/3 with =<, >=, \=
  - fixed: "Peter owns 3 kg of apples." now correctly fails
- Attempto Java Packages:
  - updated the Apache HTTP libraries to httpcore-4.1.1.jar and httpclient-4.1.1.jar
  - removed the OWL Verbalizer Webservice interface, this is now available at
    owlverbalizer.googlecode.com
  - updated enum OutputType: removed OWLRDF, added TPTP
- The APE regression test set (~3500 small ACE texts with their corresponding DRSs) is now included
  (see ape/acetext_drs.pl)
- Several improvements to the documentation of ACE, DRS and the APE source code.


ape.pl
% This file is part of the Attempto Parsing Engine (APE).
% Copyright 2008-2013, Attempto Group, University of Zurich (see http://attempto.ifi.uzh.ch).
%
% The Attempto Parsing Engine (APE) is free software: you can redistribute it and/or modify it
% under the terms of the GNU Lesser General Public License as published by the Free Software
% Foundation, either version 3 of the License, or (at your option) any later version.
%
% The Attempto Parsing Engine (APE) is distributed in the hope that it will be useful, but WITHOUT
% ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
% PURPOSE. See the GNU Lesser General Public License for more details.
%
% You should have received a copy of the GNU Lesser General Public License along with the Attempto
% Parsing Engine (APE). If not, see http://www.gnu.org/licenses/.


/* APE client for the Unix command-line.

@author Tobias Kuhn
@author Kaarel Kaljurand

Building the executable file:

==
swipl -g "working_directory(_, 'prolog/parser'), [fit_to_plp], halt."
swipl -O -F none -g "[ape], qsave_program('ape.exe', [goal(ape), toplevel(halt)])." -t halt
==

TODO:

- Do better checking of which arguments can be used together. E.g. if the user uses both -server
and -httpserver, what should ape.exe do then?

- All: Should we drop ulexreload? Otherwise all the interfaces should support it, and one would
need to add also "acetextreload" for reasons of uniformity.

- All: make timelimits configurable, currently they are hard-coded

- Socket: The socket server currently supports exactly these arguments that get_ape_results supports.
This is different from the other interfaces which also support other arguments (e.g. for fetching
files from URLs). More unification is needed.

- Java: assuming that it supports the file-parameter, how are the possible errors
communicated to Java?

- HTTP: Add logging

*/

:- consult('pack.pl').

:- use_module('prolog/ape', [
		get_ape_results/2,
		get_ape_results_timelimit/3,
		get_ape_results_timelimit/4
	]).

:- use_module('prolog/logger/error_logger').

% Default encoding used for opening files in text mode.
:- set_prolog_flag(encoding, utf8).

:- initialization on_signal(int, _, default).


%% argument(?Arg, -Value, -Desc)
%
% @tbd maybe we could describe the types more formally, so that they could be checked,
% e.g. the type of the solo value is one of [drs, drsxml, ...], which could be checked
% with memberchk/2.
%

argument('-text', '"TEXT"', 'The input ACE text. If neither -text nor -file is present then the ACE text is read from stdin.').
argument('-file', 'FILENAME', 'The name or URL of the input file containing the ACE text.').
argument('-ulextext', '"TEXT"', 'The user lexicon (taken from a string).').
argument('-ulexfile', 'FILENAME', 'The user lexicon (taken from a file or URL).').
argument('-solo', 'OUTPUT', 'Output just one output component. OUTPUT has to be one of {paraphrase,paraphrase1,paraphrase2,owlfss,owlfsspp,owlrdf,owlxml,ruleml,fol,pnf,tptp,tokens,syntax,syntaxpp,syntaxd,syntaxdpp,drs,drsxml,drspp,drshtml}.').
argument('-cinput', '', hidden).
argument('-cdrs', '', 'Output the DRS as a Prolog term.').
argument('-cdrsxml', '', 'Output the DRS in XML.').
argument('-cdrspp', '', 'Output the DRS in pretty-printed form in plain text.').
argument('-cdrshtml', '', 'Output the DRS in pretty-printed form in HTML.').
argument('-cparaphrase', '', 'Output a paraphrase which is a "best-effort" combination of paraphrase1 and paraphrase2.').
argument('-cparaphrase1', '', 'Output a paraphrase which uses full sentences instead of relative clauses.').
argument('-cparaphrase2', '', 'Output a paraphrase which uses relative clauses instead of full sentences.').
argument('-ctokens', '', 'Output tokens as a Prolog list of lists.').
argument('-csentences', '', 'Output sentences as a Prolog list.').
argument('-csyntax', '', 'Output simplified syntax trees as a Prolog list.').
argument('-csyntaxpp', '', 'Output simplified syntax trees in pretty-printed form.').
argument('-csyntaxd', '', 'Output plain syntax trees as a Prolog list (for debugging).').
argument('-csyntaxdpp', '', 'Output plain syntax trees in pretty-printed form (for debugging).').
argument('-cowlfss', '', 'Output OWL/SWRL in the Functional-Style Syntax representation (as Prolog term).').
argument('-cowlfsspp', '', 'Output OWL/SWRL in the Functional-Style Syntax representation (pretty-printed).').
argument('-cowlxml', '', 'Output OWL/SWRL in the XML representation.').
argument('-cowlrdf', '', 'Output OWL/SWRL in the RDF/XML representation. DEPRECATED').
argument('-cruleml', '', 'Output RuleML representation of the DRS.').
argument('-cfol', '', 'Output standard first-order logic representations (default form) of the DRS as a Prolog term.').
argument('-cpnf', '', 'Output standard first-order logic representations (prenex normal form) of the DRS as a Prolog term.').
argument('-ctptp', '', 'Output TPTP representation of the DRS.').
argument('-uri', 'URI', 'URI for the OWL outputs.').
argument('-noclex', '', 'Ignore the lexicon entries that are compiled into the executable.').
argument('-guess', '', 'Guess the word-class of unknown words.').
argument('-server', '', 'Launch a socket interface to APE at port 2766 (0xACE).').
argument('-httpserver', '', 'Launch an HTTP interface to APE at port 8000.').
argument('-port', 'NUMBER', 'Override the default port of either the socket or the HTTP interface.').
argument('-version', '', 'Shows version information.').
argument('-help', '', 'Shows this help page.').


%% ape is det.
%
% This is the default goal of =|ape.exe|=.
% Parses the command-line arguments, processes the arguments,
% if an ACE text is specified then parses the ACE text.
% Pretty-prints an error message in case an exception was thrown.
%
ape :-
	current_prolog_flag(argv, RawArgList),
	get_arglist(RawArgList, ArgList),
	catch(
		( arglist_namevaluelist(ArgList, InputList1), process_input(InputList1) ),
		Exception,
		format_error_for_terminal(Exception)
	).


%% get_arglist(+RawArgList, -ArgList)
%
% Returns the list of arguments.
% In SWI v6.6.0+ this can be achieved simply by:
%
%     get_arglist(ArgList) :-
%       current_prolog_flag(argv, ArgList).
%
% For backwards compatibility we assume that the argument
% list can contain '--', or the path to ape.exe before the
% first argument (which must start with '-').
%
get_arglist(RawArgList, ArgList) :-
    append(_, ['--'|ArgList], RawArgList),
    !.

% TODO: on which OS is this needed?
get_arglist([NonFlag|ArgList], ArgList) :-
	\+ atom_concat('-', _, NonFlag),
	!.

get_arglist(ArgList, ArgList).


%% process_input(+InputList:list) is det.
%
% @param InputList is a list of input parameters
%
process_input(InputList) :-
	( InputList = [] ; member(help=on, InputList) ),
	!,
	show_help.

process_input(InputList) :-
	memberchk(version=on, InputList),
	!,
	show_version.

process_input([server=on]) :-
	!,
	server.

process_input([server=on, port=Port]) :-
	!,
	server(Port).

process_input([port=Port, server=on]) :-
	!,
	server(Port).

process_input([httpserver=on]) :-
	!,
	http_server.

process_input([httpserver=on, port=Port]) :-
	!,
	http_server(Port).

process_input([port=Port, httpserver=on]) :-
	!,
	http_server(Port).

% @tbd Any other usage of server, httpserver, or port is illegal

process_input(InputList1) :-
	read_ulex(InputList1, InputList2),
	read_file(InputList2, InputList3),
	get_ape_results(InputList3, Content),
	set_utf8_encoding(user_output),
	writeln(Content).


%% show_help
%
show_help :-
	show_version,
	write('Copyright 2008-2013, Attempto Group, University of Zurich\n'),
	write('This program comes with ABSOLUTELY NO WARRANTY.\n'),
	write('This is free software, and you are welcome to redistribute it under certain conditions.\n'),
	write('Please visit http://attempto.ifi.uzh.ch for details.\n'),
	nl,
	write('Command-line arguments:\n'),
	argument(Arg, Value, Desc),
	\+ Desc = hidden,
	format('~w ~w~20|~w~n', [Arg, Value, Desc]),
	fail.

show_help.


%% show_version is det.
%
% Prints the version information.
%
show_version :-
	version(Version),
	format("Attempto Parsing Engine for ACE 6.7, version ~w~n", [Version]).


%% arglist_namevaluelist(+ArgList:list, -NameValueList:list) is det.
%
% @param ArgList is a list of arguments
% @param NameValueList is a list of ArgumentName=ArgumentValue pairs

arglist_namevaluelist([], []).

arglist_namevaluelist([Arg|Tail1], [Name=on|Tail2]) :-
    argument(Arg, '', _),
    !,
    atom_concat('-', Name, Arg),
	arglist_namevaluelist(Tail1, Tail2).

arglist_namevaluelist([Arg,ValueAtom|Tail1], [Name=Value|Tail2]) :-
    argument(Arg, _, _),
    \+ argument(ValueAtom, _, _),
    !,
    atom_concat('-', Name, Arg),
	(
		catch(atom_number(ValueAtom, ValueNumber), _, fail)
	->
		Value = ValueNumber
	;
		Value = ValueAtom
	),
	arglist_namevaluelist(Tail1, Tail2).

arglist_namevaluelist([Arg|_], _) :-
	argument(Arg, _, _),
	!,
	throw(error('Missing value for argument', context(arglist_namevaluelist/2, Arg))).

arglist_namevaluelist([Arg|_], _) :-
	throw(error('Illegal argument', context(arglist_namevaluelist/2, Arg))).


%% read_file(+InputListIn:list, -InputListOut:list) is det.
%
% @param InputListIn is a list of APE parameters
% @param InputListOut is a modified list of APE parameters
%
read_file(InputListIn, [text=AceText | InputListOut]) :-
	select(file=AceFile, InputListIn, InputListOut),
	!,
	filename_to_filecontent(AceFile, AceText).

read_file(InputList, InputList) :-
	member(text=_, InputList),
	!.

read_file(InputList, [text=AceText | InputList]) :-
	prompt(_, ''),
	read_stream_to_codes(user_input, AceTextCodes),
	atom_codes(AceText, AceTextCodes).


%% read_ulex(+InputListIn:list, -InputListOut:list) is det.
%
% Stores the user lexicon in a local (temporary) file.
% Modifies the list of APE parameters to include the name
% of the local file.
%
% @param InputListIn is a list of APE parameters
% @param InputListOut is a modified list of APE parameters
%
read_ulex(InputListIn, [ulextext=UlexText | InputListOut]) :-
	select(ulexfile=UlexFile, InputListIn, InputListOut),
	!,
	filename_to_filecontent(UlexFile, UlexText).

read_ulex(InputList, InputList).


% Note: we use: set_stream(In, encoding(utf8))
% This makes characters travel correctly through the socket.
% There might be other (better) solutions though.

:- use_module(library(streampool)).

:- style_check(-singleton).

port(2766).  % ape (0xACE)
%port(2767).  % ape-alpha (0xACF)
%port(2768).  % ape-old (0xAD0)

server :-
	port(Port),
	server(Port).

server(Port) :-
	get_time_formatted(Time),
	format(user_error, "~w: Starting a socket interface for APE at port ~w ...~n", [Time, Port]),
	tcp_socket(Socket),
	tcp_bind(Socket, Port),
	tcp_listen(Socket, 5),
	tcp_open_socket(Socket, In, _Out),
	add_stream_to_pool(In, accept(Socket)),
	stream_pool_main_loop.

accept(Socket) :-
	tcp_accept(Socket, Slave, Peer),
	tcp_open_socket(Slave, In, Out),
	set_utf8_encoding(In),
	set_utf8_encoding(Out),
	add_stream_to_pool(In, client(In, Out, Peer)).

client(In, Out, _Peer) :-
	catch(
		client_x(In, Out, _Peer),
		CatchType,
		(
			format(user_error, "~w~n", [CatchType]),
			(is_stream(In) -> close(In), delete_stream_from_pool(In) ; true),
			(is_stream(Out) -> close(Out) ; true)
		)
	).


% NB: 'APESERVERSTREAMEND' on a separate line marks the end of the stream.
% @bug should we use write_canonical/2 here?
client_x(In, Out, _Peer) :-
	% We suppress warnings of the atom being longer than 5 lines.
	% This declaration seems to have effect only here.
	style_check(-atom),
	read(In, ClientRequest),
	close(In),
	(
		ClientRequest = get(I)
	->
		get_ape_results_timelimit(I, O, 20),  % 20 seconds timelimit
		format(Out, '~w~nAPESERVERSTREAMEND~n', [O])
	;
		format(Out, 'fail~nAPESERVERSTREAMEND~n')
	),
	close(Out),
	delete_stream_from_pool(In).


/**

The HTTP interface relies on SWI-Prolog HTTP support.

@see http://www.swi-prolog.org/packages/http.html

Note that the number of workers is set to 1.
Multiple workers would share the same assert/retract space and we do not want that
because APE is not completely thread-safe.
A better solution would be to make APE thread-safe and let the user decide on
the command-line on the number of workers (because the best-performing
number depends on the number of processor cores).
Note that the SWI default is 2 workers (which seems to make sense even with single-core processors).

At Prolog prompt, stop the server by:

==
?- http_stop_server(8000, []).
==

Note that http_stop_server/2 does not make the port immediately available,
you have to wait a few (two?) minutes. Observe it with 'netstat -na'.

*/


:- use_module(library('http/thread_httpd')).
:- use_module(library('http/http_dispatch')).
:- use_module(library('http/http_parameters')).
:- use_module(library('http/http_client')).


% Configure the port.
http_port(8000).

% Configure the www root.
:- http_handler('/', ape, []).
%:- http_handler('/ape/', ape, []).

% Configure the APE webservice parameters.
% @bug unify this with the main argument description
% Note that while 'text' and 'file' are optional, at least one of them must be present.
% It is probably impossible to say this declaratively.
parameters([
	text(_,         [optional(true)]),
	file(_,         [optional(true)]),
	ulextext(_,     [optional(true)]),
	ulexfile(_,     [optional(true)]),
	noclex(_,       [oneof([on, off]), optional(true)]),
	guess(_,        [oneof([on, off]), optional(true)]),
	% ulexreload(_, [optional(true)]), % @tbd
	uri(_,          [default('http://attempto.ifi.uzh.ch/ontologies/owlswrl/test')]),
	cinput(_,       [oneof([on, off]), optional(true)]), % @bug deprecated
	cdrs(_,         [oneof([on, off]), optional(true)]),
	cdrsxml(_,      [oneof([on, off]), optional(true)]),
	cdrspp(_,       [oneof([on, off]), optional(true)]),
	cdrshtml(_,     [oneof([on, off]), optional(true)]),
	cparaphrase(_,  [oneof([on, off]), optional(true)]),
	cparaphrase1(_, [oneof([on, off]), optional(true)]),
	cparaphrase2(_, [oneof([on, off]), optional(true)]),
	ctokens(_,      [oneof([on, off]), optional(true)]),
	csentences(_,   [oneof([on, off]), optional(true)]),
	csyntax(_,      [oneof([on, off]), optional(true)]),
	csyntaxpp(_,    [oneof([on, off]), optional(true)]),
	csyntaxd(_,     [oneof([on, off]), optional(true)]),
	csyntaxdpp(_,   [oneof([on, off]), optional(true)]),
	cowlfss(_,      [oneof([on, off]), optional(true)]),
	cowlfsspp(_,    [oneof([on, off]), optional(true)]),
	cowlrdf(_,      [oneof([on, off]), optional(true)]),
	cowlxml(_,      [oneof([on, off]), optional(true)]),
	cruleml(_,      [oneof([on, off]), optional(true)]),
	cfol(_,         [oneof([on, off]), optional(true)]),
	ctptp(_,        [oneof([on, off]), optional(true)]),
	solo(_,         [oneof([drs, drsxml, drspp, drshtml,
					paraphrase, paraphrase1, paraphrase2,
					tokens, sentences,
					syntax, syntaxpp, syntaxd, syntaxdpp,
					owlfss, owlfsspp, owlrdf, owlxml,
					ruleml,
					fol, pnf, tptp]), optional(true)])
]).

http_server :-
	http_port(Port),
	http_server(Port).

http_server(Port) :-
	get_time_formatted(Time),
	format(user_error, "~w: Starting an HTTP interface for APE at port ~w ...~n", [Time, Port]),
	http_server(http_dispatch, [port(Port), workers(1)]),
	thread_get_message(_),
	halt.

%% ape(+Request) is det.
%
% This is the HTTP interface toplevel where input is received
% and results/errors are output.
%
ape(Request) :-
	parameters(Parameters),
	catch(
		(
			http_parameters(Request, Parameters),
			http_parameters_to_ape_parameters(Parameters, ApeParameters),
			get_ape_results_timelimit(ApeParameters, ContentType, Content, 10) % 10 second timelimit
		),	
		Exception,
		format_error_for_http(Exception, ContentType, Content)
	),
	format('Content-type: ~w\r\n\r\n~w', [ContentType, Content]).


%% http_parameters_to_ape_parameters(+HttpParameters:list, -ApeParameters:list) is det.
%
% Converts the parameters that have been instantiated by http_parameters/2
% into the parameters' format that APE accepts. Uninstantiated parameters are filtered out.
%
% @param HttpParameters is the result of http_parameters/2
% @param ApeParameters is the inputlist for get_ape_results_timelimit/4
%
http_parameters_to_ape_parameters([], []).

http_parameters_to_ape_parameters([Parameter | Parameters], Out) :-
	arg(1, Parameter, Arg1),
	(
		nonvar(Arg1)
	->
		functor(Parameter, Functor, _),
		key_value_to_parameter(Functor, Arg1, ApeParameter),
		Out = [ApeParameter | ApeParameters]
	;
		Out = ApeParameters
	),
	http_parameters_to_ape_parameters(Parameters, ApeParameters).


%% key_value_to_parameter(+Key:atom, +Value:atom, -ApeParameter:term) is det.
%
% Constructs an APE parameter which has a form Key = Value.
% Maps 'file' to 'text', 'ulextext' to 'ulexfile', etc.
%
% Note that for security reasons, the value of 'file' and 'ulexfile' cannot point
% to a local file (e.g. /etc/passwd). Only http-urls are allowed.
% The following example shows an illegal query.
%
%==
% http://attempto.ifi.uzh.ch:8000/?file=/etc/passwd
%==
%
% @param Key is a parameter name
% @param Value is the parameter value
% @param ApeParameter is the corresponding APE parameter
%
key_value_to_parameter(file, HttpUrl, text = FileContent) :-
	!,
	httpurl_to_filecontent(HttpUrl, FileContent).

key_value_to_parameter(ulexfile, HttpUrl, ulextext = FileContent) :-
	!,
	httpurl_to_filecontent(HttpUrl, FileContent).

key_value_to_parameter(Key, Value, Key = Value).


%% filename_to_filecontent(+FileName:atom, -FileContent:atom) is det.
%
% Reads the content of a file that is specified by FileName into an atom.
% FileName can be a URL or a regular file name. In case of URLs, only the http-prefix
% is allowed. The content of the file is expected to be encoded in UTF-8.
%
% @param FileName is the name (possibly a URL) of a file
% @param FileContent is the content (as an atom) of the file
%
filename_to_filecontent(FileName, FileContent) :-
	(
		is_http_url(FileName)
	->
		httpurl_to_filecontent(FileName, FileContent)
	;
		read_file_to_codes(FileName, Codes, [encoding(utf8)]),
		atom_codes(FileContent, Codes)
	).


%% httpurl_to_filecontent(+HttpUrl:atom, -FileContent:atom) is det.
%
% Makes an HTTP GET request to a http-resource.
%
% @param HttpUrl is a URL (expected to start as 'http://')
% @param FileContent is the content (as an atom) of the resource
% @throws socket_error (and other exceptions by http_get/3)
% @throws 'HTTP request failed' in case HTTP status code was not 'ok' (200)
%
% @bug Check if http_get expects UTF-8 or can handle other encodings as well
% @bug Make sure that http_get/3 never accesses local files (e.g. that it does not support 'file://')
%
httpurl_to_filecontent(HttpUrl, FileContent) :-
	http_get(HttpUrl, FileContent, [ user_agent('ape.exe (http://attempto.ifi.uzh.ch)'), reply_header(ReplyHeader) ]),
	memberchk(status(Code, Message), ReplyHeader),
	(
		Code = ok
	->
		true
	;
		with_output_to(atom(MessageAtom), format("~s: ~w", [Message, HttpUrl])),
		throw(error('HTTP request failed', context(httpurl_to_filecontent/2, MessageAtom)))
	).


%% is_http_url(+Atom:atom) is det.
%
% Tests is an atom is an HTTP URL.
% The test is quite naive.
%
% @param Atom is an atom to be tested.
%
% @bug Make use of SWI library url.pl
% @bug What is faster for substring matching sub_atom/5 or concat_atom/2,
%      e.g. concat_atom(['http://', _], FileName)
%
is_http_url(Atom) :-
	sub_atom(Atom, 0, 7, _, 'http://').


%% format_error_for_terminal(+Exception:term) is det.
%
% Pretty-prints the exception term for the terminal.
%
% @param Exception is the exception term, usually in the form
%        error(Formal, context(Module:Name/Arity, Message))
%
format_error_for_terminal(error(Formal, context(Predicate, Message))) :-
	!,
	format_message(Message, FMessage),
	format(user_error, "ERROR: ~w: ~w: ~w~n", [Formal, Predicate, FMessage]).

format_error_for_terminal(Error) :-
	format(user_error, "ERROR: ~w~n", [Error]).


%% format_error_for_http(+Exception:term, -ContentType:atom, -Content:term) is det.
%
% Generates an error message from the exception term.
%
% @param Exception is the exception term, usually in the form
%        error(Formal, context(Module:Name/Arity, Message))
% @param ContentType is the content type that message is formatted into, e.g. text/xml
% @param Content is the formatted error message
%
% @tbd return the same error messages as apews.perl
% @bug not XML-safe
%
format_error_for_http(error(Formal, context(Predicate, Message)), 'text/xml', Xml) :-
	!,
	functor(Formal, Name, _),
	format_message(Message, FMessage),
	with_output_to(atom(Xml), format("<error type=\"~w\">~w: ~w: ~w</error>", [Name, Formal, Predicate, FMessage])).

format_error_for_http(Error, 'text/xml', Xml) :-
	with_output_to(atom(Xml), format("<error>~w</error>", [Error])).


%% format_message(+Message:term, -FormattedMessage:atom)
%
% Formats the message that sometimes comes with the
% exception term. Often this message is unbound. The
% purpose of this rule is to return an empty atom if this
% is the case.
format_message(Message, '') :-
	var(Message),
	!.

format_message(Message, Message).


%% set_utf8_encoding(+Stream)
%
% Sets the encoding of the given stream to UTF-8. For some unknown reason, an error is sometimes
% thrown under Windows XP when calling set_stream/2, and this error is catched here.

set_utf8_encoding(Stream) :-
	catch(
		set_stream(Stream, encoding(utf8)),
		_,
		true
	).


%% get_time_formatted(-FormattedTimestamp)
%
% Generates a timestamp for the current time
%
get_time_formatted(FormattedTimestamp) :-
	get_time(Timestamp),
	format_time(atom(FormattedTimestamp), '%F %T%z', Timestamp).
